java-frontend/src/main/java/org/sonar/java/UCFGJavaVisitor.java

```diff
@@ -0,0 +1,372 @@
+
+    getAnnotatedStringParameters(methodTree).forEach(parameter -> buildBlockForParameter(parameter, blockBuilder, idGenerator));
+
+    Label nextBlockLabel = UCFGBuilder.createLabel(Integer.toString(cfg.entry().id()));
+    blockBuilder.jumpTo(nextBlockLabel);
+    return blockBuilder;
+  }
+
+  private void buildBlockForParameter(VariableTree parameter, BlockBuilder blockBuilder, IdentifierGenerator idGenerator) {
+    Expression.Variable parameterVariable = variableWithId(idGenerator.lookupIdFor(parameter.symbol()));
+    List<AnnotationTree> annotationList = parameter.modifiers().annotations();
+    List<Expression> annotationVariables = new ArrayList<>();
+
+    annotationList.forEach(annotationTree -> {
+      Expression.Variable var = variableWithId(idGenerator.newIdFor(annotationTree));
+      annotationVariables.add(var);
+      blockBuilder.assignTo(var, call(annotationTree.annotationType().symbolType().fullyQualifiedName()).withArgs(parameterVariable), location(annotationTree));
+    });
+
+    Expression[] args = annotationVariables.toArray(new Expression[annotationVariables.size()]);
+    blockBuilder.assignTo(parameterVariable, call("__annotation").withArgs(args), location(parameter.simpleName()));
+  }
+
+  private static Stream<VariableTree> getAnnotatedStringParameters(MethodTree methodTree) {
+    return methodTree.parameters().stream().filter(parameter -> isString(parameter.type().symbolType())).filter(parameter -> !parameter.modifiers().annotations().isEmpty());
+  }
+
+  private UCFGBuilder.BlockBuilder buildBasicBlock(CFG.Block javaBlock, MethodTree methodTree, IdentifierGenerator idGenerator) {
+    UCFGBuilder.BlockBuilder blockBuilder = UCFGBuilder.newBasicBlock(String.valueOf(javaBlock.id()), location(javaBlock));
+
+    javaBlock.elements().forEach(e -> buildCall(e, blockBuilder, idGenerator));
+
+    Tree terminator = javaBlock.terminator();
+    if (terminator != null && terminator.is(Tree.Kind.RETURN_STATEMENT)) {
+      ExpressionTree returnedExpression = ((ReturnStatementTree) terminator).expression();
+      Expression retExpr = constant(IdentifierGenerator.CONST);
+      if(isString(methodTree.returnType().symbolType())) {
+        retExpr = idGenerator.lookupExpressionFor(returnedExpression);
+      }
+      blockBuilder.ret(retExpr, location(terminator));
+      return blockBuilder;
+    }
+
+    Set<CFG.Block> successors = javaBlock.successors();
+    if (!successors.isEmpty()) {
+      blockBuilder.jumpTo(successors.stream().map(b -> UCFGBuilder.createLabel(Integer.toString(b.id()))).toArray(Label[]::new));
+      return blockBuilder;
+    }
+    Preconditions.checkState(javaBlock.id() == 0);
+    blockBuilder.ret(constant("implicit return"), location(methodTree.lastToken()));
+    return blockBuilder;
+  }
+
+  private void buildCall(Tree element, UCFGBuilder.BlockBuilder blockBuilder, IdentifierGenerator idGenerator) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([squid:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=1952&issues=AWJOrIvL2-ApEdl5qAxZ&open=AWJOrIvL2-ApEdl5qAxZ)

[ISSUE_KEY]: AWJOrIvL2-ApEdl5qAxZ